focus/operations/src/pull.rs

```diff
@@ -0,0 +1,210 @@
+// Copyright 2022 Twitter, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::{bail, Context, Result};
+use focus_util::app::{App, ExitCode};
+use focus_util::git_helper;
+use std::path::{Path, PathBuf};
+use std::sync::Arc;
+
+const FOCUS_SYNC_REF: &str = "refs/focus/sync";
+// This is correct for `source`
+const PREFETCH_DEFAULT_REF: &str = "refs/prefetch/remotes/origin/master";
+
+/// Entry point for pull operation
+/// Run preflight checks: ref existence, prefetch and focus/sync refs are ahead of `HEAD`
+/// Update refs/remotes/origin/* from refs/prefetch/*, and
+/// Update current branch from focus/sync
+#[tracing::instrument]
+pub fn run(app: Arc<App>, repo_path: PathBuf) -> Result<ExitCode> {
+    preflight_checks(app.clone(), repo_path.as_path())?;
+    update_refs_from_prefetch(app.clone(), repo_path.as_path())?;
+    update_current_branch(app, repo_path.as_path())?;
+
+    Ok(ExitCode(0))
+}
+
+/// Validate existence of refs that `focus pull` operates on
+fn validation_ref_existence(app: Arc<App>, repo_path: &Path) -> Result<()> {
+    let focus_sync_ref = git_helper::parse_ref(app.clone(), repo_path, FOCUS_SYNC_REF);
+    let prefetch_default_ref = git_helper::parse_ref(app, repo_path, PREFETCH_DEFAULT_REF);
+
+    if focus_sync_ref.is_err() || focus_sync_ref.unwrap().is_empty() {
+        bail!("Could not find focus sync ref or ref is empty");
+    }
+    if prefetch_default_ref.is_err() || prefetch_default_ref.unwrap().is_empty() {
+        bail!("Could not find prefetch ref or ref is empty");
+    }
+
+    Ok(())
+}
+
+/// Validate that `prefetch` refs are ahead of current HEAD, and
```

Comment:
even safer would probably have been to make sure that focus/sync is newer than head, and that pre-fetch is newer or equal to focus/sync.

Reply:
`make sure that focus/sync is newer than head`-  That does happen later or are you saying that it should happen first?

`pre-fetch is newer or equal to focus/sync` - why should focus pull care about that? As long as it's newer than HEAD, it's useful work. The background sync or something else should probably worry about that imo

Reply:
I would say that 'focus pull'-ing to a commit that diverges from pre-fetch would be un-intended behavior. You could spend a shit ton of time rebasing onto focus/sync just to realize that it contained pii, or an amended commit, and now you have to walk back and rebase onto the new origin/master. Instead, we could catch the divergence before doing the first rebase, and just do the second. It's pretty race-condition-y though.